src/arithmetic/list_funcs/list_funcs.c

```diff
@@ -761,6 +761,478 @@ SIValue AR_REDUCE
 	return accum;
 }
 
+static inline XXH64_hash_t hash_sivalue(SIValue val) {
+	// initialize the hash state
+	XXH64_state_t state;
+	XXH_errorcode res = XXH64_reset(&state, 0);
+	ASSERT(res != XXH_ERROR);
+
+	SIValue_HashUpdate(val, &state);
+
+	// finalize the hash
+	return XXH64_digest(&state);
+}
+
+// list : sorted list
+// return a list with no duplicate values
+static SIValue _dedupSorted(const SIValue list) {
+	uint32_t n   = SIArray_Length(list);
+	SIValue  res = SI_Array(n);
+
+	SIValue prev = SIArray_Get(list, 0);
+	SIArray_Append(&res, prev);
+
+	for(uint i = 1; i < n; i++) {
+		SIValue cur = SIArray_Get(list, i);
+
+		if(SIValue_Compare(prev, cur, NULL) != 0) {
+			SIArray_Append(&res, cur);
+			prev = cur;
+		}
+	}
+
+	return res;
+}
+
+static dict *_construct_histogram(const SIValue *list, uint len) {
+	dict *values_a = HashTableCreate(&dt);
+
+	for(uint i = 0; i < len; i++) {
+		SIValue val = SIArray_Get(*list, i);
+		XXH64_hash_t hash = hash_sivalue(val);
+
+		uint64_t *count = HashTableFetchValueULL(values_a, (void *)hash);
+		if(count == NULL) {
+			// first time we see this value, set count to 1
+			int ret = HashTableAddULL(values_a, (void *)hash, 1);
+			ASSERT(ret == DICT_OK);
+		} else {
+			(*count)++;
+		}
+	}
+
+	return values_a;
+}
+
+static SIValue handle_union_bag_semantic_dup_2(const SIValue *_A, const SIValue *_B) {
+	SIValue A = *_A;
+	SIValue B = *_B;
+	uint a_len = SIArray_Length(A);
+	uint b_len = SIArray_Length(B);
+	SIValue res = SI_Array(a_len + b_len);
+
+	dict *values_a = _construct_histogram(_A, a_len);
+	dict *values_b = _construct_histogram(_B, b_len);
+
+	// append values from A
+	for(uint i = 0; i < a_len; i++) {
+		SIValue val = SIArray_Get(A, i);
+		XXH64_hash_t hash = hash_sivalue(val);
+
+		uint64_t *count_a = HashTableFetchValueULL(values_a, (void *)hash);
+		uint64_t *count_b = HashTableFetchValueULL(values_b, (void *)hash);
+
+		if(count_b) {
+			*count_a = MAX(*count_a, *count_b);
+			*count_b = 0;
+		}
+
+		if(*count_a > 0) {
```

Comment:
not sure if this if condition is needed. I'm expecting `count_a` to always be > 0

Reply:
@swilly22 Aren't you intending to delete this function?

Reply:
@swilly22 I think ``count_a`` always will be > 0. I removed that condition in [8485912](https://github.com/RedisGraph/RedisGraph/pull/2908/commits/8485912916b34dc61916a5e2ff78f8a5218ba817)